File: Cpp/Medium/451. 根据字符出现频率排序.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> map1;
        for(auto c : s) {
            map1[c]++;
        }
        vector<string> bucket(s.size() + 1);
        for(auto kv : map1) {
            bucket[kv.second].push_back(kv.first);
        }
        string res;
        for(int n = bucket.size() - 1; n >= 0; n--) {
            for(auto c : bucket[n]) {
                res += string(n, c);
            }
        }
        return res;
    }
};

File: Java/medium/451. 根据字符出现频率排序.java
class Solution {
    public String frequencySort(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        for(char c : s.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }
        List<Character>[] bucket = new ArrayList[s.length() + 1];
        for(char c : map.keySet()) {
            int f = map.get(c);
            if(bucket[f] == null) {
                bucket[f] = new ArrayList<>();
            }
            bucket[f].add(c);
        }
        StringBuilder str = new StringBuilder();
        for(int i = bucket.length - 1; i >= 0; i--) {
            if(bucket[i] == null) {
                continue;
            }
            for(char c : bucket[i]) {
                for(int j = 0; j < i; j++) {
                    str.append(c);
                }
            }
        }
        return str.toString();
    }
}
